BitwardenShared/Core/Tools/Repositories/ImportCiphersRepository.swift

```diff
@@ -0,0 +1,184 @@
+import AuthenticationServices
+import BitwardenSdk
+import Foundation
+
+/// A protocol for a `ImportCiphersRepository` which manages importing credentials needed by the UI layer.
+///
+protocol ImportCiphersRepository: AnyObject {
+    /// Performs an API request to import ciphers in the vault.
+    /// - Parameters:
+    ///   - credentialImportToken: The token used in `ASCredentialImportManager` to get the credentials to import.
+    ///   - progressDelegate: Delegate to update progress.
+    /// - Returns: A dictionary containing the localized cipher type (key) and count (value) of that type
+    /// that was imported, e.g. ["Passwords": 3, "Cards": 2].
+    @available(iOS 18.2, *)
+    func importCiphers(
+        credentialImportToken: UUID,
+        progressDelegate: ProgressDelegate
+    ) async throws -> [ImportedCredentialsResult]
+}
+
+// MARK: - DefaultImportCiphersRepository
+
+/// A default implementation of a `ImportCiphersRepository`.
+///
+class DefaultImportCiphersRepository {
+    // MARK: Properties
+
+    /// The service that handles common client functionality such as encryption and decryption.
+    let clientService: ClientService
+
+    /// The factory to create credential managers.
+    let credentialManagerFactory: CredentialManagerFactory
+
+    /// The service that manages importing credentials.
+    let importCiphersService: ImportCiphersService
+
+    /// The service used to handle syncing vault data with the API.
+    let syncService: SyncService
+
+    // MARK: Initialization
+
+    /// Initialize a `DefaultImportCiphersRepository`
+    ///
+    /// - Parameters:
+    ///   - clientService: The service that handles common client functionality such as encryption and decryption.
+    ///   - credentialManagerFactory: A factory to create credential managers.
+    ///   - importCiphersService: A service that manages importing credentials.
+    ///   - syncService: The service used to handle syncing vault data with the API.
+    ///
+    init(
+        clientService: ClientService,
+        credentialManagerFactory: CredentialManagerFactory,
+        importCiphersService: ImportCiphersService,
+        syncService: SyncService
+    ) {
+        self.clientService = clientService
+        self.credentialManagerFactory = credentialManagerFactory
+        self.importCiphersService = importCiphersService
+        self.syncService = syncService
+    }
+}
+
+// MARK: ImportCiphersRepository
+
+extension DefaultImportCiphersRepository: ImportCiphersRepository {
+    @available(iOS 18.2, *)
+    func importCiphers( // swiftlint:disable:this function_body_length
+        credentialImportToken: UUID,
+        progressDelegate: ProgressDelegate
```

Comment:
🎨 Since the repository doesn't need to keep a reference to this delegate and since there's only a single method in the delegate, I wonder if this could be replaced with a closure? The main benefit is that it'd keep the logic for updating the progress when calling this function inline rather than broken out into a separate function.

Reply:
I did it because I thought it was easier to test and I was planning on adding more methods and use it weakly but I can do the change to closure approach given that I haven't added any additional features to the `ProgressDelegate` and I'm only using it directly in one method.